### PR TITLE
Apply wxSVGFileDC changes to 3.0 branch

### PR DESCRIPTION
**Pull request by MaartenBent**
This PR applies all the relevant changes to `wxSVGFileDC` from master to 3.0 branch as requested by @wnitzan in https://github.com/wxWidgets/wxWidgets/pull/215#issuecomment-475730304
I also back-ported 4c8aaf9, so `wxDC` looks the same as `wxSVGFileDC`.

Some will probably break ABI/API so have to be removed, including:
- 23cba41 changes function signature
- 65abf90 type change (does this break ABI?)
- a582d99 removes functions

I hope someone can confirm this.

### CODE REVIEW

File: interface/wx/dcsvg.h [hunk 1]
```diff
@@ -21,24 +21,30 @@
...
 */
+
 class wxSVGFileDC : public wxDC
 {
 public:
     /**
         Initializes a wxSVGFileDC with the given @a f filename with the given
-        @a Width and @a Height at @a dpi resolution.
+        @a Width and @a Height at @a dpi resolution, and an optional @a title.
+        The title provides a readable name for the SVG document.
```
Comment (vadz) [inline interface/wx/dcsvg.h]: Should document title as new since 3.0.5.

File: interface/wx/dcsvg.h [hunk 2]
```diff
@@ -56,10 +62,32 @@ class wxSVGFileDC : public wxDC
...
+        @endcode
+
+        or derive your own bitmap handler class and use it if you need to
+        customize the bitmap handling further.
+
+        @param handler The new bitmap handler. If non-NULL, this object takes
+            ownership of this handler and will delete it when it is not needed
+            any more.
+
+        @since 3.1.0
```
Comment (vadz) [inline interface/wx/dcsvg.h]: 3.0.5

**Review by vadz — COMMENTED** [review]
Sorry but I'm rather afraid of applying these changes, I'm far from certain they're not breaking something. Maybe people who really need this in their code could just merge this PR changes into their own wx builds?

If not, this would require somebody to carefully run ABI version checks using the tools described in `docs/contributing/how-to-release.md`.

**Comment by MaartenBent** [comment]
Thanks for the pointers.

@wnitzan, which fixes are badly needed on the wxPython side? Just bug fixes?
Because I'll probably leave some new features out (custom bitmap handling, setting svg title).

**Comment by wnitzan** [comment]
1. SVGFileDC doesn't produce the same drawing as DC does (sometimes SVGFileDC clearly violates expected behavior, and at others just produces unexpected drawing).
2. SVGFileDC doesn't support all the Pen and Brush styles that DC does.
3. The generated SVG file might have issues too. I'm only able to view it in a web browser, but not any other SVG viewer or editor program that I tried.

**Comment by MaartenBent** [comment]
Thanks, these are not related to the new features so I'll leave those out.

File: include/wx/dcsvg.h [hunk 3]
```diff
@@ -25,15 +27,49 @@
 
 class WXDLLIMPEXP_FWD_BASE wxFileOutputStream;
 
+class WXDLLIMPEXP_FWD_CORE wxSVGFileDC;
 
+// Base class for bitmap handlers used by wxSVGFileDC, used by the standard
+// "embed" and "link" handlers below but can also be used to create a custom
+// handler.
+class WXDLLIMPEXP_CORE wxSVGBitmapHandler
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply include/wx/dcsvg.h]: Added for the new functions (but with `30005`).

File: include/wx/dcsvg.h [hunk 4]
```diff
@@ -25,15 +27,49 @@
...
+                               wxCoord x, wxCoord y,
+                               wxOutputStream& stream) const;
+};
 
 class WXDLLIMPEXP_CORE wxSVGFileDCImpl : public wxDCImpl
 {
 public:
     wxSVGFileDCImpl( wxSVGFileDC *owner, const wxString &filename,
-                     int width=320, int height=240, double dpi=72.0 );
+                     int width = 320, int height = 240, double dpi = 72.0,
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply include/wx/dcsvg.h]: I will not back-port this feature.

File: interface/wx/dcsvg.h (hunk 1 above)
Comment (MaartenBent) [reply interface/wx/dcsvg.h]: I will not back-port this feature.

File: interface/wx/dcsvg.h (hunk 2 above)
Comment (MaartenBent) [reply interface/wx/dcsvg.h]: I will not back-port this feature.

File: include/wx/dcsvg.h [hunk 5]
```diff
@@ -48,33 +84,18 @@ class WXDLLIMPEXP_CORE wxSVGFileDCImpl : public wxDCImpl
         return -1;
     }
 
-    virtual void Clear()
-    {
-        wxFAIL_MSG(wxT("wxSVGFILEDC::Clear() Call not implemented \nNot sensible for an output file?"));
-    }
+    virtual void Clear();
```
[the remark this answers is not in the document]
Comment (MaartenBent) [reply include/wx/dcsvg.h]: I'll test it with the ABI tools. It used to throw asserts so it is probably not implemented widely.

**Comment by MaartenBent** [comment]
I tested with `abi-compliance-checker`, it shows no problems: [compat_report.zip](https://github.com/wxWidgets/wxWidgets/files/3026127/compat_report.zip)
Not all new functions are shown, I think because they are header only. If I implement them in source they do show up.

I also tested running the drawing sample in Windows with the 'save as SVG' feature back-ported. Everything keeps working when running with the new DLL (both rls and dbg).

Ready to merge, I suppose.

I added another commit which fixes annoying stc documentation regeneration. Recent typo fixes were not applied to the stc script generating it.

**Comment by vadz** [comment]
Thanks a lot for your work and testing!

The changes to wxSTC now look fine to me, but -- and I'm sorry for being such a bore -- now I'm not sure about the elliptic arc changes because formally speaking they are incompatible and if people had workarounds in their code exchanging angles for some platform, this is going to break them. So for maximum ease of mind, I think it would be better to not apply those. So can I apply the changes to wxSVG only, and not those to wxDC?

**Comment by MaartenBent** [comment]
Sure, you can skip the first commit. It should not affect wxSVG.

I back-ported it to make comparing the svg file to the wxDC easier, but it could indeed interfere with workarounds.

**Comment by vadz** [comment]
Thanks again Maarten for your (as usual excellent) work!

**Comment by wnitzan** [comment]
Thank you for the patch. It is part of wxPython 4.0.6.
But, I notice that DrawArc and DrawEllipticArc always show a solid line, even when a different pen style is selected. Is that a known issue in wxWidgets?

**Comment by MaartenBent** [comment]
It was not known (as far as I know), thanks for reporting. The pen pattern is indeed not used, other style elements (colour, width) should be fine (it works in the c++ svg sample). I'll fix it for 3.0 and 3.1.

`DrawRectangle`, `DrawPolygon` and `DrawEllipse` also don't seem use the pen pattern, so those might be affected as well. I'll compare the svg output to the screen output to see if they need fixing.

Next time please report on https://trac.wxwidgets.org. That way, other people will be able to search for, and find, your bug report. And more developers (not only participants of this PR) will see it / get notified.

**Comment by wnitzan** [comment]
I apologize. I don't feel comfortable posting an official bug report, since I only use the library via wxPython. I can't be sure, that the bug is in the library.

I would like to mention another issue (I didn't see it reported):
DrawText() and DrawRotatedText() show text in SVG (but not in DC) with extra space between characters, as if the font is monospace, although it isn't (e.g. Calibri).

Thank you for your support.

**Comment by MaartenBent** [comment]
I checked the (wx3.0) svg sample and it looks fine to me.
Maybe you see the effects of the [textLength](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/textLength) attribute ([#17271](https://trac.wxwidgets.org/ticket/17271)). This is needed so text starts and ends at the same position compared to wxDC.

svg sample:
svg sample
saved as svg (viewed in chrome 76):
svg text

**Comment by wnitzan** [comment]
Interesting.
Here's how I set the font and draw in wxPython code:

font = wx.Font(18, wx.FONTFAMILY_DEFAULT, wx.FONTSTYLE_NORMAL, wx.FONTWEIGHT_NORMAL, False, "Verdana")
dc.SetFont(font)
dc.DrawRotatedText("Verdana at 0 deg.", 100,  100, 0)

On DC:
print_to_DC

On SVG (captured and saved as 
print_to_SVG
:

**Comment by MaartenBent** [comment]
Do you do anything else to the DC?

Can you draw a box around the text and check if the DC and svg looks the same:
```c++
wxFont font = wxFont(18, wxFONTFAMILY_DEFAULT, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL, false, "Verdana");
dc.SetFont(font);
wxString s = "Verdana at 0 deg.";
wxSize textSize = dc.GetTextExtent(s);
wxPoint p = wxPoint(100, 100);
dc.DrawRectangle(p, textSize);
dc.DrawRotatedText(s, p, 0);
```

And maybe show the relevant xml from the svg (`<g...><recct>...<text>...</g>`).

**Comment by wnitzan** [comment]
I drew a yellow rectangle around the text.

Here's the DC output:
draw_text_DC

Here's the SVG output:
draw_text_SVG

Here's the SVG file (renamed with .txt):
[draw_text.txt](https://github.com/wxWidgets/wxWidgets/files/3459747/draw_text.txt)

Here's the wxPython program (renamed with .txt):
[py_draw_text.txt](https://github.com/wxWidgets/wxWidgets/files/3459748/py_draw_text.txt)

**Comment by MaartenBent** [comment]
I think I found the problem. Your screen DPI is >100%? 

I can reproduce it with c++ when I use a different DPI. Text in an SVG only has `font-size` and is DPI independent. But when using `GetTextExtent`, it uses a `wxScreenDC` to determine the size that the text should occupy (so it looks the same as the DC output in the application). This size is larger than the actual text, In addition, `textLength` causes the text to be stretched out.

I'll fix it by scaling the entire text attribute.

**Comment by wnitzan** [comment]
Indeed, I'm at 125%, on Win7.
But, I reduced DPI to 100%, and the SVG image still shows too much space between characters. It looks as if the overall text "box" is the same size as in the DC application, but each character is slimmer, taking less space, thus showing excess spacing.

**Comment by wnitzan** [comment]
If I may, 2 more issues I see in my program with DC vs. SVG.

1. Brush fill style BDIAGONAL_HATCH is not implemented, but FDIAGONAL_HATCH is...
2. Stroke overlap:
When I draw one line over another, of the same width of 1 pixel, and with opaque colors (as the default is), "ghosting" appears along the line's edges in the SVG output, whereas in DC the second line cleanly conceals the first. It might just be how SVG draws...?

In my example, I draw 2 horizontal white lines, upper and lower.
Then, I draw a black line over the lower white line.
In DC, only the upper white line remains visible. Not so in SVG.
I use this technique in my program to open gaps in already drawn lines.

DC output:
draw_line_over_line_DC

SVG output:
draw_line_over_line_SVG

SVG file (renamed with .txt):
[draw_line_over_line.txt](https://github.com/wxWidgets/wxWidgets/files/3465346/draw_line_over_line.txt)

Python file (renamed with .txt):
[py_draw_line_over_line.txt](https://github.com/wxWidgets/wxWidgets/files/3465347/py_draw_line_over_line.txt)

**Comment by MaartenBent** [comment]
Thanks for reporting. For the text-size issue I have created https://github.com/wxWidgets/wxWidgets/pull/1466. I'll port it to 3.0 as well.

I'll see if I can create patterns for the missing brush styles.

The ghosting is an effect of anti-aliasing. It can be fixed by adding [shape-rendering: crispEdges;](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/shape-rendering). This might look good for straight horizontal and vertical lines or rectangles, but any other shape will look ugly (in my opinion, see the circle in the link for example).

`wxDC` does not use anti-aliasing but a `wxDC` with `wxGraphicsContext` does use it, and it has the same ghosting problem. `wxGraphicsContext` has an option to turn it off (`SetAntialiasMode`). I'll try to add something similar to `wxSVGFileDC`, i.e. `SetShapeRenderingMode`.

**Comment by wnitzan** [comment]
I was able to turn off anti-aliasing when drawing line-over-line by adding
shape-rendering="crispEdges"
to each line in the SVG file. See attached (renamed .txt).
[draw_line_over_line_no_anti_alias.txt](https://github.com/wxWidgets/wxWidgets/files/3474697/draw_line_over_line_no_anti_alias.txt)
The bottom line is now completely gone, at any magnification ratio.

So, what's needed is a way to turn anti-aliasing on and off at will, before and after drawing something in SVGFileDC (I still need anti-aliasing for everything else I draw...).

**Comment by MaartenBent** [comment]
> I was able to turn off anti-aliasing when drawing line-over-line by adding
> shape-rendering="crispEdges"
> to each line in the SVG file. See attached (renamed .txt).
> [draw_line_over_line_no_anti_alias.txt](https://github.com/wxWidgets/wxWidgets/files/3474697/draw_line_over_line_no_anti_alias.txt)
> The bottom line is now completely gone, at any magnification ratio.
> 
> So, what's needed is a way to turn anti-aliasing on and off at will, before and after drawing something in SVGFileDC (I still need anti-aliasing for everything else I draw...).

A normal `wxDC` (without `wxGraphicsContext`) does not have anti-aliasing. Otherwise you would have the problems from the bottom line in your application as well.

I proposed to add `SetShapeRenderingMode` to set the mode. In c++ it could be used like the example below. You can temporarily change the rendering mode, draw something with that mode, then set the default mode.

```c++
void MyCanvas::Draw(wxDC& dc)
{
    [drawing anti-aliased]

    if (wxSVGFileDC* svgdc = wxDynamicCast(&dc, wxSVGFileDC)) {
        svgdc->SetShapeRenderingMode(wxSVG_SHAPE_RENDERING_CRISP_EDGES);
    }

    [draw aliased]

    if (wxSVGFileDC* svgdc = wxDynamicCast(&dc, wxSVGFileDC)) {
        svgdc->SetShapeRenderingMode(wxSVG_SHAPE_RENDERING_AUTO);
    }

    [drawing anti-aliased]
}
```

**Comment by wnitzan** [comment]
Using a transparent pen to draw a rectangle or a polygon is mistaken for a transparent brush, resulting nothing being shown (no outline nor fill).

**Comment by MaartenBent** [comment]
I tried the following code, and it works fine for me. Please provide an example that shows the problem.

```c++
dc.SetPen(*wxTRANSPARENT_PEN);
dc.SetBrush(*wxGREEN_BRUSH);
dc.DrawRectangle(100, 100, 200, 200);

wxPoint star[5];
star[0] = wxPoint(100, 60);
star[1] = wxPoint(60, 150);
star[2] = wxPoint(160, 100);
star[3] = wxPoint(40, 100);
star[4] = wxPoint(140, 150);
dc.SetBrush(*wxRED_BRUSH);
dc.DrawPolygon(WXSIZEOF(star), star, 100, 100);
```

**Comment by wnitzan** [comment]
I'm using the library through wxPython, so perhaps its the binding.

This is my Python code:
(rename to .py)
[transparent_pen_issue_py.txt](https://github.com/wxWidgets/wxWidgets/files/4271070/transparent_pen_issue_py.txt)

The generated SVG file is:
(rename to .svg)
[transparent_pen_issue_svg.txt](https://github.com/wxWidgets/wxWidgets/files/4271071/transparent_pen_issue_svg.txt)

**Comment by MaartenBent** [comment]
Thanks, I see the problem. The diagonal hatch is drawn using the transparent pen.

**Comment by wnitzan** [comment]
Thanks for confirming.
Both, forward and backward hatch are drawn like that.

**Pull request merged**